firebase-sessions/src/test/kotlin/com/google/firebase/sessions/RemoteSettingsTest.kt

```diff
@@ -0,0 +1,185 @@
+    fakeFetcher.responseJSONObject = fetchedResponse
+    remoteSettings.updateSettings()
+
+    assertThat(remoteSettings.sessionEnabled).isNull()
+    assertThat(remoteSettings.samplingRate).isEqualTo(0.75)
+    assertThat(remoteSettings.sessionRestartTimeout).isEqualTo(40.minutes)
+
+    remoteSettings.clearCachedSettings()
+  }
+
+  @Test
+  fun RemoteSettings_successfulRefetchUpdatesCache() {
+    val firebaseApp = FakeFirebaseApp().firebaseApp
+    val context = firebaseApp.applicationContext
+    val firebaseInstallations = FakeFirebaseInstallations("FaKeFiD")
+    val fakeFetcher = FakeRemoteConfigFetcher()
+
+    val remoteSettings =
+      RemoteSettings(
+        context,
+        firebaseInstallations,
+        SessionEvents.getApplicationInfo(firebaseApp),
+        fakeFetcher,
+        SESSION_TEST_CONFIGS_NAME
+      )
+    val fetchedResponse = JSONObject(validResponse)
+    fetchedResponse.getJSONObject("app_quality").put("cache_duration", 1)
+    fakeFetcher.responseJSONObject = fetchedResponse
+    remoteSettings.updateSettings()
+
+    assertThat(remoteSettings.sessionEnabled).isFalse()
+    assertThat(remoteSettings.samplingRate).isEqualTo(0.75)
+    assertThat(remoteSettings.sessionRestartTimeout).isEqualTo(40.minutes)
+
+    fetchedResponse.getJSONObject("app_quality").put("sessions_enabled", true)
+    fetchedResponse.getJSONObject("app_quality").put("sampling_rate", 0.25)
+    fetchedResponse.getJSONObject("app_quality").put("session_timeout_seconds", 1200)
+
+    // Sleep for a second before updating configs
+    Thread.sleep(2000)
+
+    fakeFetcher.responseJSONObject = fetchedResponse
+    remoteSettings.updateSettings()
+
+    assertThat(remoteSettings.sessionEnabled).isTrue()
+    assertThat(remoteSettings.samplingRate).isEqualTo(0.25)
+    assertThat(remoteSettings.sessionRestartTimeout).isEqualTo(20.minutes)
+
+    remoteSettings.clearCachedSettings()
+  }
+
+  @Test
+  fun RemoteSettings_successfulFetchWithEmptyConfigRetainsOldConfigs() {
+    val firebaseApp = FakeFirebaseApp().firebaseApp
+    val context = firebaseApp.applicationContext
+    val firebaseInstallations = FakeFirebaseInstallations("FaKeFiD")
+    val fakeFetcher = FakeRemoteConfigFetcher()
+
+    val remoteSettings =
+      RemoteSettings(
+        context,
+        firebaseInstallations,
+        SessionEvents.getApplicationInfo(firebaseApp),
+        fakeFetcher,
+        SESSION_TEST_CONFIGS_NAME
+      )
+    val fetchedResponse = JSONObject(validResponse)
+    fetchedResponse.getJSONObject("app_quality").put("cache_duration", 1)
+    fakeFetcher.responseJSONObject = fetchedResponse
+    remoteSettings.updateSettings()
+
+    assertThat(remoteSettings.sessionEnabled).isFalse()
+    assertThat(remoteSettings.samplingRate).isEqualTo(0.75)
+    assertThat(remoteSettings.sessionRestartTimeout).isEqualTo(40.minutes)
+
+    fetchedResponse.remove("app_quality")
+
+    // Sleep for a second before updating configs
+    Thread.sleep(2000)
```

Comment:
Do the sleeps slow down tests?

Reply:
Ah I see - we should be able to avoid sleeps. Sleeps are bad cause they slow down the tests and are flakey. We get the time here it seems: https://github.com/firebase/firebase-android-sdk/blob/8c12472a74bf95b35df3fbf29928420a106ba6f9/firebase-sessions/src/main/kotlin/com/google/firebase/sessions/settings/SettingsCache.kt#L73

We can pass this in as an function parameter to RemoteSettings. That way we can fast forward time in the test. We do that in our Session Initiator Tests: https://github.com/firebase/firebase-android-sdk/blob/8c12472a74bf95b35df3fbf29928420a106ba6f9/firebase-sessions/src/test/kotlin/com/google/firebase/sessions/SessionInitiatorTest.kt#L136

Reply:
Tried making the refactor to accomodate the TimeProvider, but that seem to take a lot more refactor that what it is today. Example: Time class internally does not provide time comparison with another time value. Until that is done, it is hard to refactor to use that class. So, will leave it as is for now since the time internal to sleep here is just 2 seconds.